dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpression.java

```diff
@@ -914,4 +916,85 @@ private static boolean isVarArgsInvocation(
     return TypesUtils.getArrayDepth(ElementUtils.getType(lastParamElt))
         != TypesUtils.getArrayDepth(lastArgType);
   }
+
+  /**
+   * Find the declaration of the receiver of a method call in a method tree.
+   *
+   * <p>The receiver should appear in one of two places, either as a formal parameter to the method,
+   * or as a local variable.
+   *
+   * @param tree the method tree
+   * @param receiver the receiver for which to look up a declaration
+   * @return the declaration of the receiver of the method call, if found. Otherwise, null
+   */
+  public static @Nullable VariableTree getReceiverDeclarationInMethod(
+      @Nullable MethodTree tree, JavaExpression receiver) {
+    if (tree == null) {
+      return null;
+    }
+    List<? extends VariableTree> params = tree.getParameters();
+    for (VariableTree param : params) {
+      if (param.getName().toString().equals(receiver.toString())) {
+        return param;
+      }
+    }
+    for (StatementTree statement : tree.getBody().getStatements()) {
```

Comment:
I think this only scans over the top-level statements in the method, but does not check sub-statements, where the variable might also be defined.

Maybe ask Suzanne about how to implement this method.

Reply:
@smillst Do you have any suggestions on how to implement the search such that it looks at the entire method body, including sub-statements? The specification for the method is [here](https://github.com/typetools/checker-framework/pull/6685/files/dddf1a81f415eb48548855b370c2bbaca4b6978a#r1657334302) so I won't repeat it here.

Reply:
You could use a visitor over the body.

Reply:
That is true, but I am implementing this method in the `JavaExpression` utility class, which I do not think has the capabilities of the other visitor classes existing in the Checker Framework.

Reply:
`JavaExpression` contains a `visit` class.  Also see `JavaExpressionVisitor`.  So `JavaExpression` is *supposed* to support the visitor pattern (whether it does or not).

Reply:
You should implement a subclass of TreeScanner.  (A scanner is a TreeVisitor that visits all the child tree nodes.)

Reply:
Per my comment above, I'm not sure the JavaExpression has enough information to disambiguate between two declarations of variables with the same name.  But if that doesn't occur (as determined by a full pass, or at least a full enough pass to find two), then it would be possible to choose the single declaration.  Or, do we have a TreePath to the occurrence of the receiver?  That would be best, I think, because it guarantees uniqueness.